### PR TITLE
Optimize notes and tracking workflow

### LINKED ISSUE #9409: Finalize usability optimizations of notes and tracking information
## Description

Follow up of #9399 

Some open issues were found during implementation and testing of #9399. These are summarized in this issue.
Please note that each issue can be solved independently of the others.

## Expected Behavior

1. Use the same red highlighting and crossed out text styling when hovering over the 'delete' icon as on product items.
2. When creating a new back office order (on page admin/orders/new) do not display the 'notes' field until a first item has been added to the order (hide it just like tracking information)
image
3. Display 'delete' icon only when note or tracking information is present.
image
4. Add a confirmation dialog: Are you sure? when deleting a note or tracking information. Use the same styling as when deleting a product from the order.
image
5. Do not show the 'Tracking information:' row without any further information in the shipping email when there was tracking information provided which got deleted later.
image

## Actual Behaviour

1. There is not red highlighting or crossed our text styling when hovering over the 'delete' icon.
2. Note is always visible, but table looks broken (see screenshot above)
3. Delete icon is always present.
4. There is no confirmation dialog.
5. If there was a tracking information added, which then has been deleted, in the shipping email there will be the 'Tracking information:' row without any further information

## Steps to Reproduce

1. Go to admin/orders
2. Click the 'edit' icon next to an order
3. Check the behaviour of notes and tracking information.

## Animated Gif/Screenshot

See above.

## Severity
bug-s4: it's annoying, but you can use it

## Your Environment

* Version used: staged [6ad0749ce8fbc31fda4adedd4d571d0d5fa8e733](https://github.com/openfoodfoundation/openfoodnetwork/commit/6ad0749ce8fbc31fda4adedd4d571d0d5fa8e733)
* Browser name and version: Firefox 102.0 (64-bit)
* Operating System and version (desktop or mobile): Desktop Windows 10

### PR DESCRIPTION
**Pull request by binarygit**
#### What? Why?
Closes #9409

This PR picks up after #9399 incorporating a requirement that was omitted in addition to making other changes as outlined in #9409  

#### What should we test?

- Visit admin/orders/order_num/edit
- Ascertain that an alert box shows up when clicking delete for notes and tracking info
- Ascertain clicking on 'Ok' deletes the note/tracking info
- Ascertain clicking on 'Cancel' does not delete the note/tracking info

- Ascertain that there's no delete link for tracking info/notes when user has not created a note/tracking info
---
- Ascertain that the tracking info row in the shipping email is not present when the tracking info attribute is blank
---
- Ascertain that on admin/orders/edit the note field is not visible until an item has been created.

#### Release notes

Changelog Category: User facing changes

The title of the pull request will be included in the release notes.
ezgif com-gif-maker

### CODE REVIEW

**Comment by binarygit** [comment]
so, I had some time yesterday and refactored shipments.js...let me know if it's not wanted and I'll just revert the changes. Also, I know OFN is trasitioning to `StimulusJs`. I could start the process of converting these components into Stimulus if you guys want.

File: app/views/spree/shipment_mailer/shipped_email.html.haml [hunk 1]
```diff
@@ -14,7 +14,7 @@
     = item.variant.options_text
     %br
 
-- if @shipment.tracking
+- if @shipment.tracking && !@shipment.tracking.empty?
```
Comment (jibees) [inline app/views/spree/shipment_mailer/shipped_email.html.haml:17]: Could we use 

```ruby
if @shipment.tracking.present?
```

File: app/views/spree/admin/orders/_form.html.haml [hunk 2]
```diff
@@ -6,7 +6,8 @@
     = render 'spree/admin/orders/insufficient_stock_lines', insufficient_stock_lines: @order.insufficient_stock_lines
 
   = render :partial => "spree/admin/orders/shipment", :collection => @order.shipments, :locals => { :order => order }
-  = render partial: "spree/admin/orders/note", locals: { order: @order }
+  - unless @order.shipments.empty?
```
Comment (jibees) [inline app/views/spree/admin/orders/_form.html.haml:9]: I think we should better use 
```
if @order.shipments.present?
```
instead

Also, I'm wondering if this was a requirement of the original issue ?

Comment (binarygit) [reply app/views/spree/admin/orders/_form.html.haml:9]: Updated both places with `present?`. It makes for cleaner code, thanks!
And yeah, this is a requirement, it satisfies num. 2 on https://github.com/openfoodfoundation/openfoodnetwork/issues/9409

Comment (binarygit) [reply app/views/spree/admin/orders/_form.html.haml]: Also I'll get started on shifting all this to StimulusJs but I think I should work on it at a separate PR. So I will create an issue for that.

Comment (jibees) [reply app/views/spree/admin/orders/_form.html.haml]: Ok, found that point `#2`, thanks for pointing me that out. 
Event it's written, `hide it just like tracking information`, I don't think we should test for `@order.shipments` but more `order.line_items` or `@line_items` right? I think Note can be relevant on order that don't have any shipment. 

And as it's a requirement, I think we should add it into the `What should we test?` section. 

> Also I'll get started on shifting all this to StimulusJs but I think I should work on it at a separate PR. So I will create an issue for that.

I agree. Feel free to open an issue for that.

Comment (binarygit) [reply app/views/spree/admin/orders/_form.html.haml]: ah! I see what you mean. Just didn't realize that there was a model `LineItem` particularly for this scenario. I could create a local variable called line items and pass it into the `form` partial and then we could just use `if line_items render..this`, I would just need to add this to the 'admin/orders/edit' view:
```ruby
    %div{"data-hook" => "admin_order_edit_form"}
    # The line_items local var is the addition here
      = render :partial => 'form', :locals => { :order => @order, :line_items => @order.line_items.exists?  }
 ```

File: app/assets/javascripts/admin/spree/orders/shipments.js.erb [hunk 3]
```diff
@@ -39,87 +39,79 @@ $(document).ready(function() {
...
+  }
+
+  deleteTrackingInfo = function(){
+    let shipmentNumber = $(this).data('shipment-number');
+    let tracking = ''
+
+    confirmDelete(getTrackingUrl(shipmentNumber), { shipment: { tracking: tracking } })
+  }
+
+  getTrackingUrl = function(shipmentNumber){
```
Comment (mkllnk) [inline app/assets/javascripts/admin/spree/orders/shipments.js.erb:63]: I would have called it `trackingUrl` to not confuse it with the HTTP GET we use here, too. We are simply assembling a string without "getting" it from somewhere.

**Review by mkllnk — APPROVED** [review]
Thank you! This is a great improvement.

File: app/assets/javascripts/admin/spree/orders/shipments.js.erb (hunk 3 above)
Comment (binarygit) [reply app/assets/javascripts/admin/spree/orders/shipments.js.erb]: I understand what you mean and I agree so I changed it to `trackingUrl`

**Comment by drummer83** [comment]
Thank you for another pull request @binarygit!

I did a test, here are my results regarding the requirements of the issue:

1. Use the same red highlighting and crossed out text styling when hovering over the 'delete' icon as on product items. **--> SOLVED**
2. When creating a new back office order (on page admin/orders/new) do not display the 'notes' field until a first item has been added to the order (hide it just like tracking information) **--> SOLVED**
3. Display 'delete' icon only when note or tracking information is present. **--> SOLVED**
4. Add a confirmation dialog: Are you sure? when deleting a note or tracking information. Use the same styling as when deleting a product from the order. **--> SOLVED**
5. Do not show the 'Tracking information:' row without any further information in the shipping email when there was tracking information provided which got deleted later. **--> SOLVED**

1, 3 and 4 can be seen in this screen capture:
Peek 2022-07-21 13-52

2 can be seen in this screen capture:
Peek 2022-07-21 13-59

5 can be seen in this screenshot (left before PR, right after PR):
grafik

I can also confirm that all requirements of your **What should we test?** section above are working as expected.

I have tested as an enterprise user and did a quick check as a super admin user well - all looking good!

This one is ready to go! Thanks again! :rocket:

**Pull request merged**